File: lua/perfect-numbers/perfect-numbers.lua
local function aliquot_sum(n)
  sum = 0  
  for i=1, n-1, 1 do
    if n % i == 0 then sum = sum + i end
  end
  return sum
end

local function classify(n)
  local aliquot_sum = aliquot_sum(n)
  if aliquot_sum == n then return "perfect" end
  if aliquot_sum > n then return "abundant" end
  if aliquot_sum < n then return "deficient" end   
end

return {
  aliquot_sum = aliquot_sum,
  classify = classify
}

File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify' :: Int -> Maybe Classification
classify' x
  | x < 1 = Nothing
  | comparison == LT = Just Deficient
  | comparison == EQ = Just Perfect
  | comparison == GT = Just Abundant
    where
      aliquotSum = sum [ y | y <- [1..x-1], mod x y == 0]
      comparison = compare aliquotSum x
    
classify''' :: Int -> Maybe Classification
classify''' x
  | x < 1 = Nothing
  | otherwise = case (compare aliquotSum x) of
      LT -> Just Deficient
      EQ -> Just Perfect
      GT -> Just Abundant
  where
    aliquotSum = sum [y | y <- [1..x-1], mod x y == 0]
  
classify'' :: Int -> Maybe Classification
classify'' x = 
  case x < 1 of
    True  -> Nothing
    False -> case compare aliquotSum x of 
      LT -> Just Deficient
      EQ -> Just Perfect
      GT -> Just Abundant
  where
    aliquotSum = sum [y | y <- [1..x-1], mod x y == 0]

classify = classify''